File: src/spot.rs
//! Spot instance termination monitor.
//!
//! Polls IMDS for spot termination notices and triggers graceful shutdown
//! when a termination is imminent. AWS provides a 2-minute warning before
//! spot instance termination.

use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

use crossbeam::channel::Sender;
use log::{debug, info, warn};

use crate::aws::imds::ImdsClient;
use crate::service::SupervisorBase;

/// Default polling interval for spot termination notices.
const POLL_INTERVAL: Duration = Duration::from_secs(5);

/// IMDS path for spot instance action (termination/stop notices).
const SPOT_INSTANCE_ACTION_PATH: &str = "spot/instance-action";

/// Starts the spot termination monitor in a background thread.
///
/// The monitor polls IMDS every 5 seconds for spot termination notices.
/// When a termination notice is detected, it triggers a graceful shutdown
/// via the supervisor.
pub fn start_spot_termination_monitor(
    imds_client: ImdsClient,
    base_ref: Arc<Mutex<SupervisorBase>>,
    timeout_tx: Sender<()>,
) {
    thread::spawn(move || {
        debug!(
            "Starting spot termination monitor (polling every {:?})",
            POLL_INTERVAL
        );
        monitor_loop(imds_client, base_ref, timeout_tx);
    });
}

/// Main monitoring loop that polls IMDS for spot termination notices.
fn monitor_loop(
    imds_client: ImdsClient,
    base_ref: Arc<Mutex<SupervisorBase>>,
    timeout_tx: Sender<()>,
) {
    loop {
        thread::sleep(POLL_INTERVAL);

        match check_spot_termination(&imds_client) {
            Ok(Some(action)) => {
                info!(
                    "Spot termination notice received: action={}, time={}",
                    action.action, action.time
                );
                info!("Initiating graceful shutdown due to spot termination");
                base_ref.lock().unwrap().stop(timeout_tx);
                return;
            }
            Ok(None) => {
                // No termination scheduled, continue polling.
            }
            Err(e) => {
                // Log warning but continue polling - could be transient network issue
                warn!("Failed to check spot termination status: {}", e);
            }
        }
    }
}

/// Spot instance action details returned by IMDS.
#[derive(Debug)]
struct SpotAction {
    action: String,
    time: String,
}

/// Checks IMDS for a spot termination notice.
///
/// Returns:
/// - `Ok(Some(SpotAction))` if termination is scheduled
/// - `Ok(None)` if no termination is scheduled (404 from IMDS)
/// - `Err` if there was an error querying IMDS
fn check_spot_termination(imds_client: &ImdsClient) -> anyhow::Result<Option<SpotAction>> {
    match imds_client.get_metadata(SPOT_INSTANCE_ACTION_PATH) {
        Ok(response) => {
            // Parse the JSON response: {"action": "terminate", "time": "2024-01-15T12:00:00Z"}
            let response_str: &str = response.as_ref();
            let parsed: serde_json::Value = serde_json::from_str(response_str)
                .map_err(|e| anyhow::anyhow!("failed to parse spot action response: {}", e))?;

            let action = parsed["action"].as_str().unwrap_or("unknown").to_string();
            let time = parsed["time"].as_str().unwrap_or("unknown").to_string();

            Ok(Some(SpotAction { action, time }))
        }
        Err(e) => {
            // Check if it's a 404 (no termination scheduled) vs actual error
            let err_str = e.to_string();
            if err_str.contains("404") {
                Ok(None)
            } else {
                Err(e)
            }
        }
    }
}

File: src/spot.zig
//! Spot instance termination monitor.
//!
//! Polls IMDS for spot termination notices and triggers graceful shutdown
//! when a termination is imminent. AWS provides a 2-minute warning before
//! spot instance termination.

const std = @import("std");

const aws = @import("aws");

const service = @import("service.zig");

const scoped_log = std.log.scoped(.spot);

/// Default polling interval for spot termination notices (5 seconds).
const POLL_INTERVAL_NS: u64 = 5 * std.time.ns_per_s;

/// IMDS path for spot instance action (termination/stop notices).
const SPOT_INSTANCE_ACTION_PATH = "/latest/meta-data/spot/instance-action";

/// Starts the spot termination monitor in a background thread.
///
/// The monitor polls IMDS every 5 seconds for spot termination notices.
/// When a termination notice is detected, it triggers a graceful shutdown
/// via the supervisor's shutdown_requested atomic.
pub fn startSpotTerminationMonitor() void {
    const thread = std.Thread.spawn(
        .{ .stack_size = 1024 * 1024 },
        monitorLoop,
        .{},
    ) catch |err| {
        scoped_log.err("failed to spawn spot termination monitor thread: {s}", .{@errorName(err)});
        return;
    };
    thread.detach();
    scoped_log.debug("spot termination monitor started (polling every 5s)", .{});
}

/// Main monitoring loop that polls IMDS for spot termination notices.
/// Creates its own IMDS client because this runs in a detached thread that may
/// outlive the caller's aws_ctx. Sharing the main IMDS client would risk
/// use-after-free when the main thread cleans up.
fn monitorLoop() void {
    const allocator = std.heap.page_allocator;
    var imds_client = aws.ImdsClient.init(allocator, .{}) catch |err| {
        scoped_log.err("failed to initialize IMDS client for spot monitor: {s}", .{@errorName(err)});
        return;
    };
    defer imds_client.deinit();

    while (true) {
        std.Thread.sleep(POLL_INTERVAL_NS);

        if (service.isShutdownRequested()) {
            scoped_log.debug("shutdown already requested, stopping spot monitor", .{});
            return;
        }

        switch (checkSpotTermination(&imds_client)) {
            .termination_scheduled => {
                scoped_log.info("initiating graceful shutdown due to spot termination", .{});
                service.requestShutdown();
                return;
            },
            .no_termination => {},
            .check_error => |err_name| {
                scoped_log.warn("failed to check spot termination status: {s}", .{err_name});
            },
        }
    }
}

/// Result of checking spot termination status.
const CheckResult = union(enum) {
    termination_scheduled: void,
    no_termination: void,
    check_error: []const u8,
};

fn checkSpotTermination(imds_client: *aws.ImdsClient) CheckResult {
    var diagnostic: aws.imds.ServiceError = undefined;
    const response = imds_client.getMetadata(
        SPOT_INSTANCE_ACTION_PATH,
        .{ .diagnostic = &diagnostic },
    ) catch |err| {
        if (err == error.HttpError and diagnostic.httpStatus() == 404) {
            return .no_termination;
        }
        return .{ .check_error = @errorName(err) };
    };
    defer std.heap.page_allocator.free(response);

    // Parse JSON response: {"action": "terminate", "time": "2024-01-15T12:00:00Z"}
    const parsed = std.json.parseFromSlice(
        struct { action: []const u8, time: []const u8 },
        std.heap.page_allocator,
        response,
        .{ .ignore_unknown_fields = true },
    ) catch {
        return .{ .check_error = "invalid JSON response" };
    };
    defer parsed.deinit();

    // Log before deinit frees the parsed strings
    scoped_log.info("Spot termination notice received: action={s}, time={s}", .{
        parsed.value.action,
        parsed.value.time,
    });

    return .termination_scheduled;
}

test "CheckResult union works correctly" {
    const no_term: CheckResult = .no_termination;
    try std.testing.expect(no_term == .no_termination);

    const err_result: CheckResult = .{ .check_error = "test error" };
    try std.testing.expectEqualStrings("test error", err_result.check_error);

    const term: CheckResult = .termination_scheduled;
    try std.testing.expect(term == .termination_scheduled);
}
